dashboard/app/models/contact_rollups_processed.rb

```diff
@@ -250,6 +252,44 @@ def self.extract_roles(contact_data)
     uniq_roles.blank? ? {} : {roles: uniq_roles}
   end
 
+  def self.extract_state(contact_data)
+    # The priority is: school state > user state > form geo state
+    # US state in schools table is in abbreviation, must convert it back to state name.
+    school_state = extract_field_latest_value contact_data, 'dashboard.schools', 'state'
+    if school_state
+      state_name = get_us_state_from_abbr(school_state, true) || school_state
+      return {state: state_name}
+    end
+
+    user_state = extract_field_latest_value contact_data, 'dashboard.users', 'state'
+    return {state: user_state} if user_state
+
+    form_geo_state = extract_field_latest_value contact_data, 'pegasus.form_geos', 'state'
+    form_geo_state.nil? ? {} : {state: form_geo_state}
+  end
+
+  def self.extract_postal_code(contact_data)
+    # Priority: school zip > user postal code > form geo postal code
+    school_zip = extract_field_latest_value contact_data, 'dashboard.schools', 'zip'
+    return {postal_code: school_zip} if school_zip
+
+    user_postal_code = extract_field_latest_value contact_data, 'dashboard.users', 'postal_code'
+    return {postal_code: user_postal_code} if user_postal_code
+
+    form_geo_postal_code = extract_field_latest_value contact_data, 'pegasus.form_geos', 'postal_code'
+    form_geo_postal_code.nil? ? {} : {postal_code: form_geo_postal_code}
+  end
```

Comment:
Seeing this priority pattern for city, state, and postal code makes me wonder if it's time to extract a method for this.  But when I try to imagine calling such a method, it's not much cleaner:

```rb
def self.extract_postal_code(contact_data)
  postal_code = extract_field_latest_value_with_priority(
    contact_data,
    [
      ['dashboard.schools', 'zip'],
      ['dashboard.users', 'postal_code'],
      ['pegasus.form_geos', 'postal_code']
    ]
  )
  postal_code.nil? ? {} : {postal_code: postal_code}
end
```

Do you think there's another way to DRY up this code, or is it better to leave it as-is?

Reply:
My thought is to keep it as-is. My primary concern with this method is performance because it will be called for 4 million rows. Early return may not be the best pattern but it is slightly faster, easier to read and leaves room for custom processing (such as converting state abbreviation to full name if state info comes from dashboard.schools).

Btw, I messed up my PRs by merging them on top of each. The changes in `extract_state` should not be in this PR at all.